In [2]:
import matplotlib.pyplot as plt
import numpy as np

# For Loops Refresher
Remember how we learned about for loops before? 

If you're feeling a little rusty, go back to [1 - introduction.ipynb](https://hub.mybinder.org/user/swissel-computationalworkshops-03oypry1/notebooks/1%20-%20introduction.ipynb) and review for loops.

#### <span style="color:blue"> Exercise 3.1 </span>
Otherwise, here's a quick test to refresh your memory.

You have a list. Use a **for loop** to take the sum of every item in the list. 

In [1]:
my_list = [5,61,634,8, 900]

# add your for loop here



<details> <summary> <b>Click For Answer<b> </summary> 1608 </details>

### for loop sidenote!

What if we want to generate a list of incremented integers? Well that can easily be done with the range function! The default range function starts at 0 and goes up by 1 to the number **right before** the one you pass it.

For example, if you want the numbers 0, 1, 2, 3, 4, 5  ; you would use _range(6)_

Try it out below!

In [41]:
# type in
print(list(range(6)))

[0, 1, 2, 3, 4, 5]


You can change your starting value! Instead of starting at 0 (the default value), you can give the range function two arguments. The first is the starting value (inclusive) and the next is the ending value (exlusive).

In [43]:
# [2, 10)
print(list(range(2, 10)))

[2, 3, 4, 5, 6, 7, 8, 9]


In [ ]:
# make your own range list!
# try [7, 11)


<details> <summary> <b>Click For Answer<b> </summary> [7, 8, 9, 10]  </details>

The range function is especially helpful in for loops. You can itterate over a range of values, such as these here:

In [34]:
for x in range(10):
    print(x)

0
1
2
3
4
5
6
7
8
9


Lets say you don't want to type the actual length of a list in your range function. You can use the _len_ function mentioned in lecture 1! Here are a couple of examples using the range function to access items in a list:

In [22]:
# Lets print item in this list and each index
food = ['apple', 'banana', 'cheese', 'donut']

for i in range(len(food)):
    print("Index: ", i)
    print("Food: ", food[i], "\n")

Index:  0
Food:  apple 

Index:  1
Food:  banana 

Index:  2
Food:  cheese 

Index:  3
Food:  donut 



In [30]:
# We don't want the fruit! Here's how we avoid them!
food = ['apple', 'banana', 'cheese', 'donut']

for i in range(2, len(food)):
    print("Index: ", i)
    print("Food: ", food[i], "\n")

Index:  2
Food:  cheese 

Index:  3
Food:  donut 



In [29]:
# What if we want the two in the middle?
food = ['apple', 'banana', 'cheese', 'donut']

for i in range(1, 3):
    print("Index: ", i)
    print("Food: ", food[i], "\n")

Index:  1
Food:  banana 

Index:  2
Food:  cheese 



In [27]:
# Now we just want fruit!
food = ['apple', 'banana', 'cheese', 'donut']

for i in range(0, 2):
    print("Index: ", i)
    print("Food: ", food[i], "\n")

Index:  0
Food:  apple 

Index:  1
Food:  banana 



What if you have two numpy arrays, one that corresponds to the time variable, $t$, and another that corresponds to the instantaneous velocity, $v$, at a given time. The individiual points in the arrays are the instantaneuos velocity, $v_i$ in m/s, at a specific time, $t_i$ in seconds.

Calculate the final position, $x_f$ using a **for loop**, assuming that $$x_f = \Sigma_i v_i t_i + x_0$$ where $x_0$ is the initial position given in meters.

<details> <summary> <b>Click For Hint!<b> </summary> Notice that the length of the array t is the same as the length of the array v. So you can use same index to access ti and vi.  </details>

In [12]:
t = np.arange(0, 10)
v = np.array([3, 0, 4, 7, 9, 10, 5, 7, 2, 0])
x_0 = 5.
x_f = 0.

# add your for loop here.

for i in range(10):
    x_f += t[i] * v[i]
    
print(x_f + x_0)


215.0


<details> <summary> <b>Click For Answer<b> </summary> The correct answer is $x_f$ = 215 m.  </details>

Note that in the above example, we convert a list to an np.array for the $v$ array. Very often in scientific computing, np.arrays are more useful than lists, because they only contain floats or other numerical types. Because of that, you can perform mathematical operators on the *entire* array at one time. For example, we could calculate the final position without a for loop (but make sure you know how to do it with a for loop!)

In [13]:
# This is called "vector math" and is one of the many powerful tools of python!
x_f = x_0 + np.sum(v*t)
print(x_f)

215.0


## Functions
Often, you want to calculate something that you've already calculated before. You _could_ type it out _again_ or you could simply use a *function*. The advantage of functions is that it reduces the amount of code you have to write, and importantly, the amount of code you have to debug if you find that there's something wrong.

Let's write an example function called _square_.

In [14]:
def square(x):
    """returns the square of the input x"""
    return x**2

In our example, we define the function ```square``` and its input ```x``` on the first line:
```def square(x):```

The function definition always starts with ```def```, followed by the name of the function, and then the inputs are expressed inside the parentheses. The function definition ends with a colon ```:``` 

The next line ```"""returns the square of the input x"""``` is called a _docstring_ and it is a comment that tells you what the function does, what the inputs are, and what the function returns. These are useful because you can get python to tell you what the function does if you call up the docstring. Try typing ```square(shift + tab)``` in the next cell. You should see the docstring pop up below. You will need to hold down the ```shift``` and ```tab``` keys at the same time. 

You can read the docstring with the ```shift``` and ```tab``` keys for any function. Try doing it for the built-in python function ```pow()``` as well.

In [16]:
### try to read the docstring for square() and pow() here

square(2)

TypeError: square() missing 1 required positional argument: 'x'

The last line of code ```return x**2``` tells the function to what to give back to the user. In our example, if you give the function a value ```x``` it will return the square of that value. Try giving the function a value and check that it works.

In [29]:
### check that our function square works. 
### How can you use pow() to check that the function works?



#### <span style="color:blue"> Exercise 3.2 </span>

Write a function ```period``` that calculates the period, $T$, of a pendulum given an oscillation frequency, $f$, in Hz. $$ T = \frac{1}{f} $$ 

In [18]:
# define your function period here

def period(f):
    """returns the period of a pendulum given it's oscillation frequency (f)"""
    return 1 / f


Use your function to find the period for a frequency of 4 Hz.

In [20]:
# use your function here
period(4)

0.25

Sometimes we want to give a function multiple inputs. That's easy! All we have to do is add variables to the input declaration. 

In our example, we want to find the period of a pendulum given the length of the string $l$ in m and the gravitational acceleration $g$ in m/s$^2$. 

$$ f = 2\pi \sqrt{ g/l } \\ T = 1/f$$

In [38]:
def period2(grav_acc, length):
    """The period of a pendulum, given 
    the gravitational acceleration (m/s^2) and length of the pendulum"""
    
    freq = 2*np.pi * np.sqrt(grav_acc / length)
    period = 1./freq
    return period

Note that we could have solved the equations for $T$ so that our calculation would fit in one line (and it's good practice to do so!), but we wanted to show you that you can define functions with multiple lines in them. Remember to **indent the code** that is inside the function!

Try testing out our function so that you can get the same period you found above. What gravitational acceleration and length combination gives you that value?

In [61]:
# test out our new function.

period2(9.81, ((np.pi ** 2) * 9.81 * 1 / 4))

0.25

Sometimes you also want to return multiple outputs in addition to multiple inputs. For that you just need to extend the return statement to include all the values you want. 

In [62]:
def period3(grav_acc, length):
    """The period of a pendulum, given 
    the gravitational acceleration (m/s^2) and length of the pendulum"""
    
    freq = 2*np.pi * np.sqrt(grav_acc/length)
    period = 1./freq
    return period, freq

Our new function ```period3``` returns both the period and the frequency. To collect them when we call the function, we can either separate the variables with a ```,``` or we can collect them into a list.

In [64]:
# here we fill the variables my_period and my_freq 
# with the values being returned by the function period3
my_period, my_freq = period3(9.8, 9.8 * np.pi**2 /4)
print(my_period, my_freq)

0.25 4.0


In [65]:
# here we collect them into a list
rets = period3(9.8, 9.8 * np.pi**2/4)
print(rets)
my_period = rets[0]
my_freq = rets[1]
print(my_period, my_freq)

(0.25, 4.0)
0.25 4.0


In [66]:
type(rets)

tuple

#### <span style="color:blue"> Exercise 3.3 : Function Challenge </span>
Let's put it all together. 

Say you have a charge, $q$ in C, located at point $(x_s, y_s)$.

Write a function that calculates and returns the electric field, $\overrightarrow{E}$, due to the charge at a different point $(x_p, y_p)$. Since the electric field is a vector, return the magnitude, $|E|$, and x- and y-components of the field ($E_x$ and $E_y$).

The relevant equations are:
$$ d = \sqrt{ (x_p - x_s)^2 + (y_p - y_s)^2 } \\
   \theta =\tan^{-1}( \frac{y_p - y_s} {x_p - x_s} )\\
   |E| = k \frac{ q }{d^2}\\
   E_x = |E| \cos{\theta}\\
   E_y = |E| \sin{\theta}
  $$
  
  where Coloumb's constant $k = 9\times10^{9}$ N m$^2$/C$^{2}$.

In [78]:
# write your function here.
# first figure out how many inputs you have
# then how many outputs you have. 
#

def electric_field(x_p, x_s, y_p, y_s, q):
    k = 9 * 10 ** 9
    d = np.sqrt((x_p - x_s) ** 2 + (y_p - y_s) ** 2)
    theta = np.arcsin(np.radians((y_p - y_s) / (x_p - x_s)))
    E = k * (q / d ** 2)
    E_x = np.absolute(E)*np.cos(theta)
    E_y = np.absolute(E)*np.sin(theta)
    return E, E_x, E_y


In [83]:
x = electric_field(5.2 * 10 ** -2, 3.3 * 10 ** -2, -4.1 * 10 ** -2, 7.2 * 10 ** -2, 5*10**-9)
x

(3427.265803503427, 3408.7519503949457, -355.7541686932763)

In [ ]:
def efield(q, xs, ys, xp, yp ):
    k = 9e9
    d = np.sqrt((xp-xs)**2 + (yp-ys)**2)
    theta = np.arctan2((ys-yp),(xs-xp))
    absE = k * q  * pow(d, -2)
    Ex = absE * np.cos(theta)
    Ey = absE * np.sin(theta)
    return d, theta, absE, Ex, Ey

In [1]:
print(efield(1./9.*1e-9, 1, 0, 0, 0))
print(efield(1./9.*1e-9, 0, 1, 0, 0))
print(efield(5e-9, -4.1e-2, 7.2e-2, 5.2e-2, 3.3e-2))

NameError: name 'efield' is not defined

# Particle Cascades

We can illustrate the usefulness of loops and functions with a simple
example from physics. 

Cosmic ray electrons produce air showers when they interact in the atmosphere. The high-energy electron transfers some of its energy to other particles so that a cascade of hundreds or thousands of high speed particles are produced as shown in the figure below.

<img src="infiles/emshower.png">


## A Simple Model

Let's assume that there are three particles in a cascade: electrons, photons (light particles), and positrons (a positively charged electron).

Each electron or positron can split its energy with a photon that it radiates. That's drawn like this:
<img src="infiles/electron_vertices.png">

Each photon can make a pair of electrons and positrons. That's drawn like this:
<img src="infiles/photon_vertex.png">


### Now the particles multiply

So if we started out with one cosmic ray electron, we get an electron and a photon. 
Then the photon makes a new electron and a positron, while the electrons and positrons continue to make photons.

This continues so that we get a cascade of particles. 

<img src="infiles/heitler_model.png">

# Turning the Model into Code

Our goal will be to keep track of how many electrons, positrons, and photons we have in our cascade. 
We also want to know how much energy each of the particles has.

## Let's Code up the First Layer

First we have to chose an initial energy for the first electron, the cosmic ray electron. In this first layer, the electron will split its energy between an electron and a photon. 

In [5]:
# first layer
energy_cr = 160 # Joules, 1e21 eV

We'll use lists to keep track of the energy of electrons, positrons, and gammas that we have in the entire cascade. 

In [6]:
electrons = []
positrons = []
gammas = []

In [7]:
# energy of the CR splits into one electron and one positron
electrons_this_layer = [energy_cr/2.]
positrons_this_layer = []
gammas_this_layer    = [energy_cr/2.]

# these arrays store the particles made every layer
electrons.append(electrons_this_layer)
positrons.append(positrons_this_layer)
gammas.append(gammas_this_layer)

Each entry in the list is the particle's energy.

In [8]:
print(electrons_this_layer)
print(positrons_this_layer)
print(gammas_this_layer)
print(electrons)
print(positrons)
print(gammas)

[80.0]
[]
[80.0]
[[80.0]]
[[]]
[[80.0]]


The length of these lists tell us get the number of particles in the cascade.

In [9]:
print(len(electrons_this_layer))
print(len(positrons_this_layer))
print(len(gammas_this_layer))
print(len(electrons))
print(len(positrons))
print(len(gammas))

1
0
1
1
1
1


You might wonder why we need two arrays for each type of particle (e.g. electrons_this_layer and electrons) and why their shapes are different.

In [10]:
print(np.shape(electrons_this_layer))
print(np.shape(positrons_this_layer))
print(np.shape(gammas_this_layer))
print(np.shape(electrons))
print(np.shape(positrons))
print(np.shape(gammas))

(1,)
(0,)
(1,)
(1, 1)
(1, 0)
(1, 1)


We are using two different arrays because we want to add more particles at each layer. Each row contains a layer's worth of particles. This will be more apparent as we add more layers.

## Let's add another layer

Now all electrons make a gamma and an electron (bremstraahlung)
and all positrons make a gamma and a positron (bremstraahlung)
And all photons make a electron and positron pair (pair production)

First we'll store the last layer in a new list and then start on the new layer.

In [11]:
electrons_last_layer = electrons_this_layer
positrons_last_layer = positrons_this_layer
gammas_last_radlength = gammas_this_layer

# reset the arrays
electrons_this_layer = []
positrons_this_layer = []
gammas_this_layer    = []

Now we'll use for loops to make new particles for each particle in the previous layer. We'll start with the electrons and positrons.

In [12]:
# electrons from the last layer
# each electron makes an electron and a gamma
for el in electrons_last_layer:
    gammas_this_layer.append(el/2.)
    electrons_this_layer.append(el/2.)

# positrons from the last layer
# each positron makes a positron and a gamma
for po in positrons_last_layer:
    gammas_this_layer.append(po/2.)
    positrons_this_layer.append(po/2.)

#### <span style="color:blue"> Exercise 3.1 </span>
Now you add the gammas!

In [ ]:
# add the gammas.


...and now store all the particles in the electrons, positrons, and gammas lists.

In [13]:
# now store all the particles


Now figure out how many electrons, positrons, and gammas there are in the two layers and what their energies are.

In [1]:
# how many of each particle are there?


# Many Layers
Now we want to let the cascade develop over many layers, say 10. 

In [14]:
# now we want to loop over many layers
n_layers = 10

# reset so we are starting at the beginning
# first radiation length
energy_cr = 160 # Joules, 1e21 eV

electrons = []
positrons = []
gammas = []

# energy of the CR splits into one electron and one gamma
electrons_this_layer = [energy_cr/2.]
positrons_this_layer = []
gammas_this_layer    = [energy_cr/2.]

# these arrays store the particles made every radiation length
electrons.append(electrons_this_layer)
positrons.append(positrons_this_layer)
gammas.append(gammas_this_layer)

for i_layer in range(n_layers):
    
    # next layer
    # now all electrons make a gamma and an electron (bremstraahlung)
    # and all positrons make a gamma and a positron (bremstraahlung)
    electrons_last_layer = electrons_this_layer
    positrons_last_layer = positrons_this_layer
    gammas_last_layeere = gammas_this_layer

    # reset the arrays
    electrons_this_layer = []
    positrons_this_layer = []
    gammas_this_layer    = []

    for el in electrons_last_layer:
        gammas_this_layer.append(el/2.)
        electrons_this_layer.append(el/2.)

    for po in positrons_last_layer:
        gammas_this_layer.append(po/2.)
        positrons_this_layer.append(po/2.)

    for ga in gammas_this_layer:
        electrons_this_layer.append(ga/2.)
        positrons_this_layer.append(ga/2.)

    # now store all the particles
    electrons.append(electrons_this_layer)
    positrons.append(positrons_this_layer)
    gammas.append(gammas_this_layer)
    
    # at each layer, let us know how many particles there are
    print(i_layer, ": ", len(electrons_this_layer) + len(positrons_this_layer) + len(gammas_this_layer))

0 :  4
1 :  12
2 :  36
3 :  108
4 :  324
5 :  972
6 :  2916
7 :  8748
8 :  26244
9 :  78732


In [15]:
# how many particles are there at each step?
for i_layer in range(len(electrons)):
    print("Layer: ", i_layer)
    print("    Number of electrons: ", len(electrons[i_layer]))
    print("    Number of positrons: ", len(positrons[i_layer]))
    print("    Number of gammas: ", len(gammas[i_layer]))

Layer:  0
    Number of electrons:  1
    Number of positrons:  0
    Number of gammas:  1
Layer:  1
    Number of electrons:  2
    Number of positrons:  1
    Number of gammas:  1
Layer:  2
    Number of electrons:  5
    Number of positrons:  4
    Number of gammas:  3
Layer:  3
    Number of electrons:  14
    Number of positrons:  13
    Number of gammas:  9
Layer:  4
    Number of electrons:  41
    Number of positrons:  40
    Number of gammas:  27
Layer:  5
    Number of electrons:  122
    Number of positrons:  121
    Number of gammas:  81
Layer:  6
    Number of electrons:  365
    Number of positrons:  364
    Number of gammas:  243
Layer:  7
    Number of electrons:  1094
    Number of positrons:  1093
    Number of gammas:  729
Layer:  8
    Number of electrons:  3281
    Number of positrons:  3280
    Number of gammas:  2187
Layer:  9
    Number of electrons:  9842
    Number of positrons:  9841
    Number of gammas:  6561
Layer:  10
    Number of electrons:  29525
    N

#### <span style="color:blue"> Exercise 3.2 </span>
The number of particles is growing fast! Let's make a plot to see what the dependence is. We want to put the layer number on the x-axis and the number of electrons in that layer on the y-axis. How will you make the plot?

In [16]:
# Make a plot


Your plot may look exponential. If that's true, if we plot it on a log scale, it should look like a straight line, because $\ln(e^{-\alpha x}) = -\alpha x$. Try using the [plt.semilogy(...)](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.semilogy.html) function in matplotlib to check it out. 

In [17]:
# Make a plot with the y-axis in a log-scale.


Can you also make a plot of the energy of the electrons (or positrons or gammas) at each layer?

In [ ]:
# Make a plot of the energy of the electrons vs. layer number


# Functions make the work easier

We can simplify this by defining a function that makes particles for each layer. 

The function below takes lists of electrons, positrons, and gammas as inputs. Then for each particle, it makes the appropriate new particles and splits the energy. The function returns the next layer of particles.

In [18]:
def make_particles(electrons, positrons, gammas):
    # now all electrons make a gamma and an electron (bremstraahlung)
    # and all positrons make a gamma and a positron (bremstraahlung)
    # and all gammas make a positron and an electron
    electrons_last_layer = electrons
    positrons_last_layer = positrons
    gammas_last_layer = gammas

    # reset the arrays
    electrons = []
    positrons = []
    gammas    = []
    
    for el in electrons_last_layer:
        gammas.append(el/2.)
        electrons.append(el/2.)

    for po in positrons_last_layer:
        gammas.append(po/2.)
        positrons.append(po/2.)

    for ga in gammas_this_layer:
        electrons.append(ga/2.)
        positrons.append(ga/2.)
    
    return electrons, positrons, gammas

Now that we have the function, we can simplify our for loop. 

In [23]:
# layers
n_layers = 10

# reset so we are starting at the beginning
# first layer
energy_cr = 160 # Joules, 1e21 eV

electrons = []
positrons = []
gammas = []

# energy of the CR splits into one electron and one gamma
electrons_this_layer = [energy_cr/2.]
positrons_this_layer = []
gammas_this_layer    = [energy_cr/2.]

# these arrays store the particles made every radiation length
electrons.append(electrons_this_layer)
positrons.append(positrons_this_layer)
gammas.append(gammas_this_layer)

for i_layer in range(n_layers):
    
    # next radiation length
    print(i_layer)
    # the \ continues the line on the next point
    electrons_this_layer, \
    positrons_this_layer, \
    gammas_this_layer = make_particles(electrons_this_layer,
                                          positrons_this_layer,
                                          gammas_this_layer)

    # now store all the particles
    electrons.append(electrons_this_layer)
    positrons.append(positrons_this_layer)
    gammas.append(gammas_this_layer)

0
1
2
3
4
5
6
7
8
9


#### <span style="color:blue"> Exercise 3.3 </span>
Make the same plots you made before!

In [ ]:
# Make plots


How many layers does it take for the average energy to be reduced by 1/e?

In [24]:
# use the plots to figure it out.